Sources/NIOHPACK/HPACKDecoder.swift

```diff
@@ -0,0 +1,233 @@
+            switch value {
+            case 0:
+                self = .literal
+            default:
+                self = .indexed(index: value)
+            }
+        }
+    }
+    
+    /// The maximum size of the dynamic table. This is defined in RFC 7541
+    /// to be the sum of [name-octet-count] + [value-octet-count] + 32 for
+    /// each header it contains.
+    public internal(set) var maxDynamicTableLength: Int {
+        get { return headerTable.dynamicTableAllowedLength }
+        /* private but tests */
+        set { headerTable.dynamicTableAllowedLength = newValue }
+    }
+    
+    /// Creates a new decoder
+    ///
+    /// - Parameter maxDynamicTableSize: Maximum allowed size of the dynamic header table.
+    public init(allocator: ByteBufferAllocator, maxDynamicTableSize: Int = HPACKDecoder.maxDynamicTableSize) {
+        self.headerTable = IndexedHeaderTable(allocator: allocator, maxDynamicTableSize: maxDynamicTableSize)
+        
+        // start the decode buffer at the requested size of the dynamic header list table, or the default
+        // size, whichever is larger.
+        self.decodeBuffer = allocator.buffer(capacity: max(maxDynamicTableSize, DynamicHeaderTable.defaultSize))
+    }
+    
+    /// Reads HPACK encoded header data from a `ByteBuffer`.
+    ///
+    /// It is expected that the buffer cover only the bytes of the header list, i.e.
+    /// that this is in fact a slice containing only the payload bytes of a
+    /// `HEADERS` or `CONTINUATION` frame.
+    ///
+    /// - Parameter buffer: A byte buffer containing the encoded header bytes.
+    /// - Returns: An array of (name, value) pairs.
+    /// - Throws: HpackDecoder.Error in the event of a decode failure.
+    public mutating func decodeHeaders(from buffer: inout ByteBuffer) throws -> HPACKHeaders {
+        // take a local copy to mutate
+        var bufCopy = buffer
+        self.decodeBuffer.clear()
+        
+        var headers: [HPACKHeader] = []
+        while bufCopy.readableBytes > 0 {
+            // this will update our own `decodeBuffer` property.
+            if let header = try self.decodeHeader(from: &bufCopy) {
+                headers.append(header)
+            }
+        }
+        
+        buffer = bufCopy
+        return HPACKHeaders(buffer: self.decodeBuffer, headers: headers)
+    }
+    
+    private mutating func decodeHeader(from buffer: inout ByteBuffer) throws -> HPACKHeader? {
+        // peek at the first byte to decide how many significant bits of that byte we
+        // will need to include in our decoded integer. Some values use a one-bit prefix,
+        // some use two, or four.
+        // This one-byte read is guaranteed safe because `decodeHeaders(from:)` only calls into
+        // this function if at least one byte is available to read.
+        let initial: UInt8 = buffer.getInteger(at: buffer.readerIndex)!
+        switch initial {
+        case let x where x & 0x80 == 0x80:
```

Comment:
given that you use `0b1xxxxxxx` to document, you could just write this case (and the ones below) as 

    case let x where x & 0b1000_0000 != 0:

that might make it easier to read?